Source/controls/touch/renderers.cpp

```diff
@@ -0,0 +1,52 @@
+#include "controls/touch/renderers.h"
+#include "engine.h"
+
+namespace devilution {
+
+namespace {
+
+VirtualGamepadRenderer virtualGamepadRenderer(&virtualGamepad);
+
+} // namespace
+
+void DrawVirtualGamepad(const Surface &out)
+{
+	virtualGamepadRenderer.Render(out);
+}
+
+void VirtualGamepadRenderer::Render(const Surface &out)
+{
+	directionPadRenderer.Render(out);
+	primaryActionButtonRenderer.Render(out);
+	secondaryActionButtonRenderer.Render(out);
+	spellActionButtonRenderer.Render(out);
+	cancelButtonRenderer.Render(out);
+}
+
+void VirtualDirectionPadRenderer::Render(const Surface &out)
+{
+	auto center = virtualDirectionPad->area.position;
+	auto radius = virtualDirectionPad->area.radius;
+	int diameter = 2 * radius;
+
+	int x = center.x - radius;
+	int y = center.y - radius;
+	int width = diameter;
+	int height = diameter;
+	DrawHalfTransparentRectTo(out, x, y, width, height);

```

Comment:
`SDL_FillRect(SDL_GetWindowSurface, SDL_MapRGBA)`

Reply:
So, looks like we can't draw to `SDL_GetWindowSurface` when hardware rendering is going on, but we should be able to draw to `GetOutputSurface()` which gives us nice 32bit support for the rendering.

Reply:
>  gives us nice 32bit support for the rendering

Only if the output surface is 32-bit (not the case on some platforms). Perhaps now it's best to stick to 8-bit.

Reply:
I assumed that anything wtih multi touch display also has 32bit capability. Is that not right?

Reply:
Running in 8-bit mode can still be beneficial for performance and battery life.
There are also things like HP TouchPad that have fewer bits for color.